### ISSUE TITLE
Proposal: make `ags3` the new master and move current master to a branch `ags4`

### ISSUE THREAD

**Issue by sonneveld** [adventuregamestudio/ags#757]
Through noone's fault, AGS4 development has stalled. I've noticed that some people have tried to build `master` thinking it is the development branch for the next release.

To help reduce the confusion, I propose we reset the branch changesets so `master` contains AGS3 changes and AGS4 changes are available in a branch called `ags4`.  This should be easy enough to do with git and a force push or two. (I don't think merging would help here)

I don't know if this would break the current pull requests but it looks like they can be edited to change the base, so we could request the creator modifies the base to `master` before we accept them.

**Comment by ivan-mogilko** [comment]
I'd really wish you discuss this with every developer who has plans on take an active part in the future, like @morganwillcock and @AlanDrake for instance. Current development lacks clear goals, they are not set anywhere and everyone tries to change code according to their own idea. Changes similar to one you propose here should depend on these goals imho.

**Comment by sonneveld** [comment]
Also there are a few branches we could maybe remove, requesting that the owner maintains them in their own fork until they're ready again to be merged: feature_editor_drac, legacy, experimental

**Comment by sonneveld** [comment]
> I'd really wish you discuss this with every developer who has plans on take an active part in the future,

That's why I opened this issue, to start a discussion.

**Comment by ivan-mogilko** [comment]
> Also there are a few branches we could maybe remove, requesting that the owner maintains them in their own fork until they're ready again to be merged: feature_editor_drac, legacy, experimental

"legacy" - serves like the tag created right after we started large refactoring of the code, it's not changed.
"feature_editor_drac" - I opened issue about it earlier: #479
"experimental" - this is my old branch which I had to abandon after it became messy in my opinion. I picked few changes from there over time, like features and refactoring, and it was no longer meant for development, only for the reference.

**Comment by AlanDrake** [comment]
Well, if you guys keep stubbornly adding new features to ags3 then sure, go ahead and make it master.
Ags4 main purpose was to make you suffer less when implementing new features and give the users only the latest useful features without confusing leftovers, but I see your love for spaghetti knows no bound.
How much else have you planned for ags3? I started ags4 because development had stalled and it seemed feature complete.
Maybe we can try extrapolating the engine logic so we can make a legacy engine module that can easily instanced while kept separated.
My stance remains that game developers willing to dabble with the code do not want to deal with that pile of legacy luggage.

**Comment by ivan-mogilko** [comment]
> if you guys keep stubbornly adding new features to ags3

I was not planning to work on ags3 at all, and the reason I did was that number of people kept asking for certain features that they needed to have "soon". I could not add them to ags4, because I was afraid that could make some other of their game break if they use old script API (which is not incommon), so it added like this.

First time I was waiting for other people like you and morganw to work on ags4, which could perhaps persuade me to move there as well, but noone did, so I continued in ags3. I said I was okay to switch anytime, that's why I kept merging my work to ags4 to keep it ready.

>  but I see your love for spaghetti knows no bound.

I have no love for "spaghetti". I have a tolerance to having old code around until I can fix it.

> How much else have you planned for ags3?

Personally, I've added almost everything I planned for this release, and not planning to work further.

**Comment by sonneveld** [comment]
> How much else have you planned for ags3?
> Maybe we can try extrapolating the engine logic so we can make a legacy engine module that can easily instanced while kept separated.

I was planning on opening a roadmap issue after this. I'd like to have a minor 3.6 release that concentrates on porting to SDL2 and console. Also possibly some groundwork for extrapolating engine logic.

> but I see your love for spaghetti knows no bound.

Please lets keep it civil :)

**Comment by ivan-mogilko** [comment]
I agreed to remove all the legacy stuff from the engine starting with the next version, because that's rational, and tbh I got tired of it as well. It was very sad thing to do since I put so much effort in keeping it, but I had to acknowledge my mistake doing this.

So like I say, I was perfectly okay with switching to ags4, and then doing this if time allows:
> Maybe we can try extrapolating the engine logic so we can make a legacy engine module that can easily instanced while kept separated.

But as time passed I grew extremely exhausted because of the difficulties during development, and some irl issues. Also, project is not organized as well as it has to be, and I see this as my failure but have no more strength to fix this. There's no clear set of goals for it, which makes it tough to plan ahead and keep contributions under check (if I don't know what we are aiming for, how can I judge other people suggestions?)

So I officially leave decision making to others after completing 3.5.0.

**Comment by AlanDrake** [comment]
Sorry, my years of work-related stress *may* be leaking through my words. Thanks god I quit and I'll soon be done with it.
I know my presence is akin to that of an immaterial ghost, but since I'll go back to developing my game, I'll most certainly end up fiddling with AGS code to implement new features.
I'd rather AGS4 be a thing, so I can work without causing regressions, explosions, and casualties.

Would any of you be able to outline how to handle that engine-logic extrapolation?
This attempt must be planned on the roadmap, because these kind of changes will definitely make merges impossible.
We need to figure out that roadmap and stick to it.

**Comment by ivan-mogilko** [comment]
> Would any of you be able to outline how to handle that engine-logic extrapolation?

I think we need to begin with clarifying what belongs to "engine" and what belongs to "game logic", and whether we may need several layers of "logic".

**Comment by sonneveld** [comment]
I'll open another issue. I just wanted to discuss branching here :)

**Comment by sonneveld** [comment]
> If we could straighten that up into single procedure for update and single procedure for drawing, things would become much easier from that point.

Agreed!

**Comment by sonneveld** [comment]
> I'd rather AGS4 be a thing, so I can work without causing regressions, explosions, and casualties.

That'll only be a thing until the first proper 4.0 release, then you'll still have problems with regressions, explosions, etc with v4.0 games (unless we have a plan in place for supporting them).

**Comment by ivan-mogilko** [comment]
The problem is not in potential regressions, but simply in that legacy features conflict with newer better design.

**Comment by ivan-mogilko** [comment]
>  I know you've had reservations about working alongside legacy, but perhaps it's better if you start on top of Unity or create your own fork? 

To keep this fair there's the counter way alternative: one could finish ScummVM port of ags which may support all the legacy games: https://github.com/adventuregamestudio/scummvm/tree/ags


Regarding custom engine builds, there's a "engine caps" tag system (currently unused) allowing to distinguish custom versions:
https://github.com/adventuregamestudio/ags/blob/master/Editor/AGS.Editor/DataFileWriter.cs#L1377

**Comment by AlanDrake** [comment]
One thing is causing a regression on the current engine, another thing is causing and avalanche of regressions on all its older versions (not to add the weight of keeping everything backward compatible).
Which is why there needs to be a separation of the game engines.
But the problem is that making big changes to refactor makes everything unmergeable anyway, whatever path is chosen, at least until game-logic is separated.
I wanted to do this on AGS4 to see how it could be done, the plan of course hasn't gone so well because of lack of investment on my part and new development on AGS3 (also I didn't want to make your SDL2 branch completely unmergeable).
Ideally all the legacy stuff would have gotten later the same treatment and ended up on a AGSClassic namespace/module (or similar).

I mean, it could also be done directly on AGS3, but with the legacy code it's probably going to be more work and more difficult.
Wouldn't it be better to experiment these kind of changes on AGS4 first? As Crimson agrees having to deal with the legacy stuff is a big problem for new development.

**Comment by morganwillcock** [comment]
I wasn’t sure whether to write this in the road map ticket or this one, but I think that clearly identifying the project goal will solve both issues. Right now there is not a clear position of what the goal of the program is, and so any change that isn't a bug (because programs should work) or compatibility fix (because of a precedent for backwards compatibility) is potentially deemed as controversial or incorrect. I can make a case for what I believe the program should do here, but to try and keep things concise I think that this needs another issue opened to clearly define what this program is meant to do, and once the project goal is established that will make the roadmap a solvable problem. Once the roadmap is done I think it will be clear which branches there should be and which should be the default.

**Comment by ivan-mogilko** [comment]
> I wasn’t sure whether to write this in the road map ticket or this one, but I think that clearly identifying the project goal will solve both issues. Right now there is not a clear position of what the goal of the program is, 

I totally agree with this. No "roadmap" is a "program's goal", they come after deciding a goal, as a draft plan on how to reach one, imo.

**Comment by ivan-mogilko** [comment]
Alright, it's done, I reset master branch to "ags3", and created "ags4" where the old master head was.
Also, updated both 2 pull requests that targeted "master", they now target "ags4" for consistency.

This is unfortunate, but the reality is that there's a more active development in ags3 branch for the last 1.5 years, and multiple persons got confused with master branch not building or not being what they thought. So for the time being it will be back at ags3 again, at least until SDL2 port is merged in.

I apologize to everyone for this inconvenience.

A reminder, since this was a force push, everyone who synced with our repository recently would have to do `git fetch` and then reset local "master" branch to the head of "ags3" (as an option - delete or rename old one and checkout new one).
If you've got any changes of your own at the old "master", -  don't forget to create a branch with alternative name before doing above.
